### DISCUSSION TITLE
The behavior of Installing firmwares through composite id

### DISCUSSION THREAD

**Discussion by hugh712** [fwupd/fwupd#4921] [General]
Hi,

According to the description for installing a firmware file (in a cab file) as below:
`install FILE [DEVICE-ID|GUID]     Install a firmware file on this hardware`
seems it'll only install __a firmware file__ through specific Device-id or GUID,
but actually it will upgrade all firmwares in my **docking station**.

I found that currently both fwupdmgr and fwupdtool can be used to run this command,
but the behaviors are different, 
in fwupdmgr, it will install all firmwares due to it tries to get all possible devices through the same composite_id (The composite ID will be the same value for all parent, child and sibling devices) but fwupdtool is not.

So I am not quite sure which one is the right behavior?

**Comment by superm1** [comment]
The fwupdmgr behavior you describe is the intentional behavior specifically because of composite device handling. I think we never updated the description/documentation for it though as you point out.

It sound like a second bug if fwupdtool doesn't work that way too.

**Comment by hugh712** [comment]
@superm1 

I'm sorry, would you please provide where the description of "composite device handling" is?
And yes, I believe it's a bug for fwupdtool in installing firmware. Since it only gets device as [0]  not like fwupdmgr to get all devices via composite id [1]


[0] https://github.com/fwupd/fwupd/blob/main/src/fu-tool.c#L1254
[1] https://github.com/fwupd/fwupd/blob/main/src/fu-daemon.c#L1070

**Reply by superm1** [reply]
> I'm sorry, would you please provide where the description of "composite device handling" is?

I mean the line you identified above https://github.com/fwupd/fwupd/blob/main/src/fu-util.c#L4227 doesn't actually mention anything about composite devices.  The behavior you see for fwupdmgr is what I expected, but it's not properly made clear to anyone.  So I think that help-text just needs to be expanded to explain it will install firmware for all possible devices the CAB matches.

> And yes, I believe it's a bug for fwupdtool in installing firmware. Since it only gets device as [0] not like fwupdmgr to get all devices via composite id [1]

It looks like to me 63bc42eace wasn't ported to fu-tool is the mistake.  Maybe you can open a PR to port it.

**Comment by hugh712** [comment]
sure, I will work on the PR, thank you.
